File: ios/tests_class_sel_imp/web/DoProxy.js
var SET_SKIN_COLOR = "setSkinColor";
var BULL_CLASS = "Bull";
var CATTLE_CLASS = "Cattle";

var DoProxyProto = DoProxy.prototype;
function DoProxy()
{
	this.notFirstRun = false;
	
	this.cattle = [];//new Array[3];
	
	this.say = null;
	this.skin = null;
	
	this.setSkinColor_Func = null;
	
	this.say_Func = null;
	
	this.bullClass = null;
}

DoProxyProto.setAllVars = function setAllVars()
{
	this.cattle[0] = new Cattle;
	
	this.bullClass = window[BULL_CLASS];
	this.cattle[1] = new this.bullClass;
	this.cattle[2] = new this.bullClass;
	
	this.say = Bull.prototype.saySomething;
	this.skin = Bull.prototype[SET_SKIN_COLOR];
};

DoProxyProto.doWithCattleId_colorParam = function doWithCattleId_colorParam(aCattle, color)
{
	if(this.notFirstRun === false){
		var myName = arguments.callee.name;
		console.log("Running in the method of " + myName);
		this.notFirstRun = true;
	}
	
	var cattleParamClassName = aCattle.constructor.name;
	
	if(cattleParamClassName === BULL_CLASS || 
		cattleParamClassName === CATTLE_CLASS)
	{
		aCattle.setLegsCount(4);
		if(aCattle[this.skin.name]){
			aCattle[this.skin.name](color);
		}else{
			console.log("Hi, I am a " + cattleParamClassName + ", have not setSkinColor!");
		}
		aCattle[this.say.name]();
	}else{
		var yourClassName = aCattle.constructor.name;
		console.log("Hi, you are a " + yourClassName + ", but I like cattle or bull!");
	}
};

DoProxyProto.SELFuncs = function SELFuncs()
{
	this.doWithCattleId_colorParam(this.cattle[0], "brown");
	this.doWithCattleId_colorParam(this.cattle[1], "red");
	this.doWithCattleId_colorParam(this.cattle[2], "black");
	this.doWithCattleId_colorParam(this, "haha");
};

DoProxyProto.functionPointers = function functionPointers()
{
	this.setSkinColor_Func = this.skin;//this.cattle[1][this.skin.name];
	this.say_Func = this.say;//this.cattle[1][this.say.name];
	
	this.setSkinColor_Func.call(this.cattle[1], "verbos");
	
	console.log("Running as a function pointer will be more efficiency!");
	
	this.say_Func.call(this.cattle[1]);
};

File: ios/tests_class_sel_imp/tests_class_sel_imp/DoProxy.m
//
//  DoProxy.m
//  tests_class_sel_imp
//
//  Created by 裴超 on 13-5-8.
//  Copyright (c) 2013年 裴超. All rights reserved.
//

#import "DoProxy.h"
#import "Cattle.h"
#import "Bull.h"

@implementation DoProxy

firstRun = YES;

-(void) setAllVars
{
	cattle[0] = [Cattle new];
	
	bullClass = NSClassFromString(BULL_CLASS);
	cattle[1] = [bullClass new];
	cattle[2] = [bullClass new];
	
	say = @selector(saySomething);
	skin = NSSelectorFromString(SET_SKIN_COLOR);
}

-(void) doWithCattleId:(id)aCattle colorParam:(NSString *)color
{
	if (notFirstRun == NO) {
		NSString *myName = NSStringFromSelector(_cmd);
		NSLog(@"Running in the method of %@", myName);
		notFirstRun = YES;
	}
	
	NSString *cattleParamClassName = [aCattle className];
	
	if([cattleParamClassName isEqualToString:BULL_CLASS] ||
	   [cattleParamClassName isEqualToString:CATTLE_CLASS])
	{
		[aCattle setLegsCount:4];
		if([aCattle respondsToSelector:skin])
		{
			[aCattle performSelector:skin withObject:color];
		}
		else
		{
			NSLog(@"Hi, I am a %@, have not setSkinColor!", cattleParamClassName);
		}
		[aCattle performSelector:say];
	}
	else
	{
		NSString *yourClassName = [aCattle className];
		NSLog(@"Hi, you are a %@, but I like cattle or bull!", yourClassName);
	}
}


-(void)SELFuncs
{
	[self doWithCattleId:cattle[0] colorParam:@"brown"];
	[self doWithCattleId:cattle[1] colorParam:@"red"];
	[self doWithCattleId:cattle[1] colorParam:@"black"];
	[self doWithCattleId:self colorParam:@"haha"];
}

-(void)functionPointers
{
	setSkinColor_Func = (void (*)(id, SEL, NSString *))[cattle[1] methodForSelector:skin];
	//上面的语句其实等效于下面这种方法
	//IMP setSkinColor_Func = [cattle[1] methodForSelector: skin];

	say_Func = [cattle[1] methodForSelector:say];

	setSkinColor_Func(cattle[1], skin, @"verbos");

	NSLog(@"Running as a function pointer will be more efficiency!");

	say_Func(cattle[1], say);
}

@end
